### PR TITLE
cdb2api cdb2_bind_array: bind an array to a parameter name

### PR DESCRIPTION
**Pull request by adizaimi**
This patch adds ability to bind parameter array means of new cdb2api function:
```
int cdb2_bind_array(cdb2_hndl_tp *hndl, const char *varname,
                    int type, const void *varaddr[], unsigned int count, int typelen);
```
Current supported types of arrays passed in are integers, floats, doubles, and string arrays: (CDB2_INTEGER, CDB2_REAL, CDB2_CSTRING).

Example usage:
```
        int N = 24;
        int values[N];
        for (int i = 0; i < N; i++) {
            values[i] = i;
        }
        cdb2_bind_array(hndl, "myarray", CDB2_INTEGER, (const void *)values, N, sizeof(values[0]));
        execute_sql(dbhndl, "select a,b from t2 where a in carray(@myarray) order by a\n");
```
Please see testcase in this patch for more usage examples.

### CODE REVIEW

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ sqlite/ext/comdb2/tables.c @@
+ If a new system table is added, be sure to document it in docs/pages/programming/system_tables.md.
@@ sqlite/src/vdbeInt.h @@
+ If the order of members in struct Mem is changed, columnNullValue (defined in sqlite/vdbeapi.c) needs to be reordered accordingly.
@@ db/sqlinterfaces.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
```

File: cdb2api/cdb2api.c [hunk 1]
```diff
@@ -5056,6 +5056,87 @@ int cdb2_bind_index(cdb2_hndl_tp *hndl, int index, int type,
...
+    hndl->n_bindvars++;
+    hndl->bindvars = realloc(hndl->bindvars, sizeof(CDB2SQLQUERY__Bindvalue *) *
+                                                 hndl->n_bindvars);
+    CDB2SQLQUERY__Bindvalue *bindval = malloc(sizeof(CDB2SQLQUERY__Bindvalue));
+    cdb2__sqlquery__bindvalue__init(bindval);
+    bindval->type = type;
+    bindval->varname = (char *)varname;
+    switch(type) {
+    case CDB2_INTEGER:
+        if (typelen == sizeof(int32_t)) {
```
Comment (mponomar) [inline cdb2api/cdb2api.c:5078]: We should probably also allow int16_t.  Or in the other direction, we should only allow int64, so they don't incur that extra conversion.

File: cdb2api/cdb2api.c [hunk 2]
```diff
@@ -5056,6 +5056,87 @@ int cdb2_bind_index(cdb2_hndl_tp *hndl, int index, int type,
...
+            bindval->int64_arr->n_elements = count;
+        }
+        else if (typelen == sizeof(int64_t)) {
+            bindval->int64_arr = malloc(sizeof(CDB2SQLQUERY__Bindvalue__Int64array));
+            cdb2__sqlquery__bindvalue__int64array__init(bindval->int64_arr);
+            bindval->int64_arr->elements = malloc(count * sizeof(int64_t));
+            memcpy(bindval->int64_arr->elements, varaddr, count * typelen);
+            bindval->int64_arr->n_elements = count;
+        }
+    break;
```
Comment (mponomar) [inline cdb2api/cdb2api.c:5094]: Error on unknown size?

File: cdb2api/cdb2api.c [hunk 3]
```diff
@@ -5056,6 +5056,87 @@ int cdb2_bind_index(cdb2_hndl_tp *hndl, int index, int type,
...
+    case CDB2_CSTRING:
+        bindval->text_arr = malloc(sizeof(CDB2SQLQUERY__Bindvalue__Textarray));
+        cdb2__sqlquery__bindvalue__textarray__init(bindval->text_arr);
+        bindval->text_arr->elements = malloc(count * sizeof(char *));
+        //printf("AZ: ");
+        for(int i=0; i<count;i++) {
+            bindval->text_arr->elements[i] = strdup(((char**)varaddr)[i]); // TODO: check that this gets freed
+            printf("%s ", bindval->text_arr->elements[i]);
```
Comment (mponomar) [inline cdb2api/cdb2api.c:5123]: Debug trace?

File: cdb2api/cdb2api.c [hunk 4]
```diff
@@ -5056,6 +5056,87 @@ int cdb2_bind_index(cdb2_hndl_tp *hndl, int index, int type,
...
+            bindval->real64_arr->n_elements = count;
+        }
+    break;
+    case CDB2_CSTRING:
+        bindval->text_arr = malloc(sizeof(CDB2SQLQUERY__Bindvalue__Textarray));
+        cdb2__sqlquery__bindvalue__textarray__init(bindval->text_arr);
+        bindval->text_arr->elements = malloc(count * sizeof(char *));
+        //printf("AZ: ");
+        for(int i=0; i<count;i++) {
+            bindval->text_arr->elements[i] = strdup(((char**)varaddr)[i]); // TODO: check that this gets freed
```
Comment (mponomar) [inline cdb2api/cdb2api.c:5122]: One suggestion that was made internally was to get the length from the `lengths` argument if non-null for strings.

File: db/sqlinterfaces.c [hunk 5]
```diff
@@ -3790,6 +3790,29 @@ static int bind_parameters(struct reqlogger *logger, sqlite3_stmt *stmt,
...
+                flag = 3; /* CARRAY_TEXT; */ break;
+            default:
+                logmsg(LOGMSG_ERROR, "Unsupported type %d for carray_bind\n", p.type);
+                rc = -1;
+                return rc;
+            }
+
+            extern int sqlite3_carray_bind(sqlite3_stmt *pStmt, int idx, void *aData, int nData, int mFlags, void (*xDestroy)(void*)); 
+            rc = sqlite3_carray_bind(stmt, p.pos, p.u.p, p.arraylen, flag, SQLITE_STATIC);
+            printf("AZ: binding array for int, rc=%d, first element %ld\n", rc, *(int64_t*) p.u.p);
```
Comment (mponomar) [inline db/sqlinterfaces.c:3813]: Debug trace - going to assume you're still debugging this, so not adding more debug trace commends.

**Review by mponomar — COMMENTED** [review]
Just some preliminary thoughts.  Looks good sofar!

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/d9c01b9d/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>76/386 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>socksql_master_swings<br>auth_twofiles_generated<br>auth_queueodh_generated<br>auth<br>userschema<br>cdb2sql_queueodh_generated<br>comdb2sys_queueodh_generated<br>cdb2sql<br>comdb2sys<br>userschema_queueodh_generated</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/d9c01b9d/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/d9c01b9d/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2599'>Resubmit</a></p>

File: cdb2api/cdb2api.c (hunk 1 above)
Comment (adizaimi) [reply cdb2api/cdb2api.c:5078]: I can even add int8_t...not sure what users would want. We can discuss.

File: cdb2api/cdb2api.c (hunk 3 above)
Comment (adizaimi) [reply cdb2api/cdb2api.c:5123]: Yes, this is not quite done yet : )

File: cdb2api/cdb2api.c (hunk 4 above)
Comment (adizaimi) [reply cdb2api/cdb2api.c:5122]: Sure, sounds fine.

File: plugins/newsql/newsql.c [hunk 6]
```diff
@@ -1068,6 +1068,20 @@ static int newsql_param_value(struct sqlclntstate *clnt,
     int len = val->value.len;
     void *p = val->value.data;
 
+    if (val->int64_arr && val->int64_arr->n_elements) {
```
Comment (akshatsikarwar) [inline plugins/newsql/newsql.c:1071]: Can we use `OneOf` in the proto definition and use a switch case here?

File: protobuf/sqlquery.proto [hunk 7]
```diff
@@ -39,6 +39,37 @@ message CDB2_SQLQUERY {
...
+    }
+    message textarray {
+        repeated string elements = 1;
+    }
+    message blobarray {
+        repeated bytes elements = 1;
+    }
+    /*
+       message bindarray {
+       oneof array_oneof {
```
Comment (akshatsikarwar) [inline protobuf/sqlquery.proto:60]: optional oneof

File: cdb2api/cdb2api.c (hunk 4 above)
Comment (adizaimi) [reply cdb2api/cdb2api.c:5122]: Actually, that makes it behave like a blob, and after further discussion, we would not support blobs as native type. If strings are not null terminated, they won't pack correctly by protobuf.

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ sqlite/ext/comdb2/tables.c @@
+ If a new system table is added, be sure to document it in docs/pages/programming/system_tables.md.
@@ sqlite/src/vdbeInt.h @@
+ If the order of members in struct Mem is changed, columnNullValue (defined in sqlite/vdbeapi.c) needs to be reordered accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/d16da6f1/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/d16da6f1/smoketest.log'>Error &#x26A0; </a>.<br>Failed to compile.</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/d16da6f1/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/d16da6f1/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2599'>Resubmit</a></p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ sqlite/ext/comdb2/tables.c @@
+ If a new system table is added, be sure to document it in docs/pages/programming/system_tables.md.
@@ sqlite/src/vdbeInt.h @@
+ If the order of members in struct Mem is changed, columnNullValue (defined in sqlite/vdbeapi.c) needs to be reordered accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/e05102fa/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/e05102fa/smoketest.log'>Error &#x26A0; </a>.<br>Failed to compile.</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/e05102fa/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/e05102fa/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2599'>Resubmit</a></p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ sqlite/ext/comdb2/tables.c @@
+ If a new system table is added, be sure to document it in docs/pages/programming/system_tables.md.
@@ sqlite/src/vdbeInt.h @@
+ If the order of members in struct Mem is changed, columnNullValue (defined in sqlite/vdbeapi.c) needs to be reordered accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/e05102fa/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/e05102fa/robomark.txt'>Error &#x26A0; </a>.<br>Regression testing: <b>3/388 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>lostwrite [setup failure]<br>recover_deadlock<br>logfill</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/e05102fa/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/e05102fa/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2599'>Resubmit</a></p>

File: cdb2api/cdb2api.c (hunk 2 above)
Comment (adizaimi) [reply cdb2api/cdb2api.c:5101]: Added check with lower and upper bound.

File: plugins/newsql/newsql.c (hunk 6 above)
Comment (adizaimi) [reply plugins/newsql/newsql.c]: Yep. Took a while to upgrade protoc-c but working now.

File: cdb2api/cdb2api.c (hunk 1 above)
Comment (adizaimi) [reply cdb2api/cdb2api.c:5095]: only int32 and int64 have native protobuf support on the Carray bind point of calling.

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ sqlite/ext/comdb2/tables.c @@
+ If a new system table is added, be sure to document it in docs/pages/programming/system_tables.md.
@@ sqlite/src/vdbeInt.h @@
+ If the order of members in struct Mem is changed, columnNullValue (defined in sqlite/vdbeapi.c) needs to be reordered accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/428dd6b6/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: Success &#10003;.</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/428dd6b6/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/428dd6b6/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2599'>Resubmit</a></p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ sqlite/ext/comdb2/tables.c @@
+ If a new system table is added, be sure to document it in docs/pages/programming/system_tables.md.
@@ sqlite/src/vdbeInt.h @@
+ If the order of members in struct Mem is changed, columnNullValue (defined in sqlite/vdbeapi.c) needs to be reordered accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/3d991ba3/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/3d991ba3/robomark.txt'>Error &#x26A0; </a>.<br>Regression testing: Success &#10003;.</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/3d991ba3/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/3d991ba3/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2599'>Resubmit</a></p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ sqlite/ext/comdb2/tables.c @@
+ If a new system table is added, be sure to document it in docs/pages/programming/system_tables.md.
@@ sqlite/src/vdbeInt.h @@
+ If the order of members in struct Mem is changed, columnNullValue (defined in sqlite/vdbeapi.c) needs to be reordered accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/b8c3f2bf/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>5/401 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>simple_timepart_reptimeout_generated<br>sc_timepart<br>sc_timepart_logicalsc_generated<br>simple_timepart<br>datetime</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/b8c3f2bf/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/b8c3f2bf/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2599'>Resubmit</a></p>

**Review by mponomar — APPROVED** [review]
This looks great!

**Comment by adizaimi** [comment]
Thank you Mike!

**Pull request merged**